File: python/src/testivai/runner.py
"""
Report runner — invokes the @testivai/witness CLI for the compare/report
half of the pipeline (diff, tolerances, noise hint, HTML report, exit code).

Node.js is guaranteed present in playwright-python environments: Playwright
for Python bundles a Node driver. We still resolve conservatively:
  1. TESTIVAI_CLI env var — full command override (e.g. "node /x/testivai.js")
  2. `testivai` on PATH (project installed @testivai/witness)
  3. `npx --yes @testivai/witness` (zero-install fallback)
"""

from __future__ import annotations

import os
import shutil
import subprocess
from pathlib import Path
from typing import Optional


def resolve_cli() -> list[str]:
    override = os.environ.get("TESTIVAI_CLI")
    if override:
        return override.split()
    direct = shutil.which("testivai")
    if direct:
        return [direct]
    npx = shutil.which("npx")
    if npx:
        return [npx, "--yes", "@testivai/witness"]
    raise RuntimeError(
        "testivai: could not find the @testivai/witness CLI. Install Node.js "
        "and either `npm i -D @testivai/witness` or ensure `npx` is on PATH, "
        "or set TESTIVAI_CLI to the command to run."
    )


def run_report(
    project_root: Optional[Path] = None,
    *,
    fail_on_diff: bool = False,
    open_report: bool = False,
) -> int:
    """Run `testivai report` in project_root. Returns the exit code."""
    cmd = [*resolve_cli(), "report", "-q"]
    if fail_on_diff:
        cmd.append("--fail-on-diff")
    if open_report:
        cmd.append("--open")
    completed = subprocess.run(cmd, cwd=str(project_root or Path.cwd()))
    return completed.returncode

File: java/src/main/java/ai/testiv/testivai/Runner.java
package ai.testiv.testivai;

import java.io.IOException;
import java.nio.file.Path;

/**
 * Runs the compare/report half of the pipeline via the {@code
 * @testivai/witness} CLI. Node.js is present wherever playwright-java is —
 * Playwright for Java ships a Node driver.
 *
 * <p>Resolution: {@code TESTIVAI_CLI} env override → {@code npx --yes
 * @testivai/witness}.
 */
public final class Runner {

  private Runner() {}

  /** Run {@code testivai report} in the working directory. */
  public static int runReport() {
    return runReport(Path.of("").toAbsolutePath(), false);
  }

  /** Run {@code testivai report} in {@code projectRoot}. Returns the exit code. */
  public static int runReport(Path projectRoot, boolean failOnDiff) {
    ProcessBuilder pb;
    String override = System.getenv("TESTIVAI_CLI");
    if (override != null && !override.isBlank()) {
      String[] parts = (override + " report -q" + (failOnDiff ? " --fail-on-diff" : "")).split("\\s+");
      pb = new ProcessBuilder(parts);
    } else if (failOnDiff) {
      pb = new ProcessBuilder("npx", "--yes", "@testivai/witness", "report", "-q", "--fail-on-diff");
    } else {
      pb = new ProcessBuilder("npx", "--yes", "@testivai/witness", "report", "-q");
    }
    pb.directory(projectRoot.toFile());
    pb.inheritIO();
    try {
      return pb.start().waitFor();
    } catch (IOException e) {
      System.err.println(
          "[testivai] could not run the @testivai/witness CLI (" + e.getMessage() + "). "
              + "Captures are in .testivai/temp/ — run `npx testivai report` manually.");
      return -1;
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      return -1;
    }
  }
}
